### PR TITLE
make the plugin incompatibility tests less deterministic

### PR DESCRIPTION
**Pull request by jsvd**
upgrade_pipeline has the nasty side effect of..upgrading the pipeline. This test checks that upgrade_pipeline is called and, since the configuration changed ONCE, upgrade_pipeline should be called ONCE.

When I did the test I relaxed this concern because the expectation of calling a method mocks it. Which means the original isn't called, which means the pipeline isn't upgraded so, in next cycle of reloading, the upgrade is tried again.

By ensuring the method is actually called and performs its side effect, we can deterministically test that upgrade_pipeline should be called once and only once.

fixes https://github.com/elastic/logstash/issues/5160

### CODE REVIEW

**Comment by purbon** [comment]
@jsvd have you tried running this, let's say for N times?

File: logstash-core/spec/logstash/agent_spec.rb [hunk 1]
```diff
@@ -150,10 +150,10 @@
         context "with a config that does not contain reload incompatible plugins" do
           let(:second_pipeline_config) { "input { generator { } } filter { } output { }" }
 
-          it "does not reload if new config contains reload incompatible plugins" do
+          it "does upgrade the new config" do
             t = Thread.new { subject.execute }
             sleep 0.01 until subject.running_pipelines? && subject.pipelines.values.first.ready?
-            expect(subject).to receive(:upgrade_pipeline).at_least(2).times
+            expect(subject).to receive(:upgrade_pipeline).once.and_call_original
```
Comment (purbon) [inline logstash-core/spec/logstash/agent_spec.rb]: I wonder why this change of expectations here?, do you remember why you initially started with `at_least(2)` and now require only once? do you have a reason behind?

Comment (jsvd) [inline logstash-core/spec/logstash/agent_spec.rb]: `upgrade_pipeline` has the nasty side effect of..upgrading the pipeline. This test checks that upgrade_pipeline is called, and, since the configuration changed ONCE, `upgrade_pipeline` should be called ONCE.

When I did the test I relaxed this concern because the expectation of calling a method mocks it. Which means the original isn't called, which means the pipeline isn't upgraded so, in next cycle of reloading, the upgrade is tried again.

By ensuring the method is actually called and performs its side effect, we can deterministically test that upgrade_pipeline should be called once and only once.

**Comment by jsvd** [comment]
@purbon well I can say I tried it more than N times (for some N).

**Comment by purbon** [comment]
❤️ https://github.com/elastic/logstash/pull/5175#issuecomment-213426970 ❤️

File: logstash-core/spec/logstash/agent_spec.rb (hunk 1 above)
Comment (purbon) [inline logstash-core/spec/logstash/agent_spec.rb]: makes sense to me.

**Comment by ph** [comment]
LGTM, have tested it

**Comment by purbon** [comment]
LGTM, :shipit: :shipit:
